keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,1067 @@
+    # ...other rule related fields here...
+```
+
+Pros:
+
+- Self-contained resource
+- Higher level policy abstractions can be defined out-of-tree and can be
+  translated into this form by controllers
+
+Cons:
+
+- The resource grows by the number of match rules and amounts of configuration data.
+- Need to define how configurations of overlapping matches are interpreted. If
+  any of the matches fail does the rule as an admission check fail, or if any
+  matches pass does the admission check as a whole pass?
+
+##### Configurability Option 2: Put configuration in separate resources
+
+```
+apiVersion: admissionregistration.k8s.io/v1
+kind: ValidatingAdmissionConfiguration
+metadata:
+  name: "validate-xyz.example.com"
+spec:
+  configurationCustomResourceDefinition:
+    group: rules.example.com
+    kind: LimitRule
+    ...
+      openAPIV3Schema: ... # the below configuration is validated against this schema
+  validations:
+    - rule: "self.spec.xyz == configuration.x"
+      # ...other rule related fields here...
+    ...
+---
+apiVersion: rules.example.com/v1
+kind: LimitRule
+metadata:
+  name: "validate-xyz-rule1.example.com"
+spec:
+  match:
+  - rules: ...
+    objectSelectors: …
+    namespaceSelectors ... 
+  x: 1 
+```
+
+Pros:
+
+- Aligns closely with how policy frameworks separate policy definition from
+  policy configuration
+- Scalable beyond the maximum size of a single K8s resource
+- Easy to identify/modify/track individual policies
+- More easily allows separation-of-concerns when policy is a joint effort
+  between multiple people/teams
+- Allows for updating the abstract policy definition separate from the concrete
+  policy
+
+Cons:
+
+- Significantly more complex implementation
+  - CRDs need to be created when a ValidatingAdmissionConfiguration is created
+  - Requires the API server establish watches on the CRs to accumulate all the
+    configuration information
+- Possible to implement using Option 1. A controller can construct a
+  resource for the API in Option 1 from custom resources that define this
+  option (e.g. via a controller).
+- CRD lifecycle prevents CRs from being written until the CRD is written, which
+  means that until a ValidatingAdmissionConfiguration is written AND a CRD is
+  created for it, no CRs can be written.  Applying multiple resources
+  (e.g. kubectl apply -f <lots of yaml files>) only works if the apply
+  operations are ordered and objects are verified to exist at specific steps.
+
+#### Additional Capabilities (WIP)
+
+In addition to the high level API structure proposed above, there are quite
+a few capatiblities that are important to this enhancement that we are working
+on defining. These are all work-in-progress.
+
+Rule Message:
+
+- message formatting (need support for string formatting. Allow a CEL expression
+  to be used to format the whole string? Or allow CEL expressions to evaluate
+  template values?)
+- severity (e.g. "Warning")
```

Comment:
Assuming rules can be configured to populate https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/1693-warnings via AdmissionResponse rather than reject a request, is there also a use case for another class of warning that isn't exposed to clients?

Reply:
> is there also a use case for another class of warning that isn't exposed to clients?

That sounds like audit annotations, which can be returned by admission webhooks today.

I think the info included in validating admission webhook responses covers this:
* warnings
* audit annotations
* admission decision, one of:
  * accept
  * reject (optionally with a custom status code, reason, message, details, etc)

Reply:
(note specifically that warnings and audit annotations can be returned for allowed requests)